src/client/src/Stream.c

```diff
@@ -1836,6 +1842,167 @@ STATUS setNalAdaptationFlags(PKinesisVideoStream pKinesisVideoStream, UINT32 nal
+            if(firstEvent) {
+                CHK_STATUS(mkvgenGenerateTagsChain(NULL, KVSEVENT_IMAGE_GENERATION_STRING, "", &packagedSize, MKV_TREE_TAGS));
+                CHK_STATUS(createSerializedMetadata(KVSEVENT_IMAGE_GENERATION_STRING, "", FALSE, packagedSize, event, MKV_TREE_TAGS, &serializedNodes[neededNodes++]));
+            }
+            else {
+                CHK_STATUS(mkvgenGenerateTagsChain(NULL, KVSEVENT_IMAGE_GENERATION_STRING, "", &packagedSize, MKV_TREE_SIMPLE));
+                CHK_STATUS(createSerializedMetadata(KVSEVENT_IMAGE_GENERATION_STRING, "", FALSE, packagedSize, event, MKV_TREE_SIMPLE, &serializedNodes[neededNodes++]));
+            }
+            if(hasMetadata && pMetadata->imagePrefix != NULL) {
+                CHK_STATUS(mkvgenGenerateTagsChain(NULL, KVSEVENT_IMAGE_PREFIX_STRING, pMetadata->imagePrefix, &packagedSize, MKV_TREE_SIMPLE));
+                CHK_STATUS(createSerializedMetadata(KVSEVENT_IMAGE_PREFIX_STRING, pMetadata->imagePrefix, FALSE, packagedSize, event, MKV_TREE_SIMPLE, &serializedNodes[neededNodes++]));
+            }
+            break;
+        default:
+            break;
+    }
+    // Check that custom data can fit as well
+    if(hasMetadata) {
+        for(UINT8 list = 0; list < pMetadata->numberOfPairs; list++) {
+            CHK_STATUS(mkvgenGenerateTagsChain(NULL, pMetadata->names[list], pMetadata->values[list], packagedSize, MKV_TREE_SIMPLE));
+            CHK_STATUS(createSerializedMetadata(pMetadata->names[list], pMetadata->values[list], FALSE, packagedSize, event, MKV_TREE_SIMPLE, &serializedNodes[neededNodes++]));
+        }
+    }
+    creatingNodes = FALSE;
+
+
+    // if we're the first event, then we need to track the location of event 'tags' element in the queue for potential follow-up events in this fragment.
+    if(firstEvent) {
+        pKinesisVideoStream->metadataTracker.eventIndex = metadataQueueSize;
+    }
+
+
+    //append/insert all metadata nodes to the queue.
+    if(firstEvent) {
+        for(UINT8 iter = 0; iter < neededNodes; iter++) {
+            CHK_STATUS(appendValidatedMetadata(pKinesisVideoStream, serializedNodes[iter]));
+        }
+    }
+    else {
+        ++pKinesisVideoStream->metadataTracker.eventIndex;
+
+        for(UINT8 iter = 0; iter < neededNodes; iter++) {
+            CHK_STATUS(insertValidatedMetadata(pKinesisVideoStream, serializedNodes[iter], pKinesisVideoStream->metadataTracker.eventIndex + iter));
+        }
+    }
+
+    // Unlock the stream (even though it will be unlocked in the cleanup
+    pKinesisVideoClient->clientCallbacks.unlockMutexFn(pKinesisVideoClient->clientCallbacks.customData, pKinesisVideoStream->base.lock);
+    streamLocked = FALSE;
+
+CleanUp:
+
+    //in the event we failed when creating the nodes, free the rest.
+    if(creatingNodes) {
+        for(UINT8 iter = 0; iter < neededNodes; iter++) {
+            if (serializedNodes[iter] != NULL) {
+                MEMFREE(serializedNodes[iter]);
```

Comment:
We should call the safe version, `SAFE_MEMFREE` here to NULL check before freeing and set to NULL when done.